read_terragrunt_config

yorinasub17 opened the pull request:
This implements `read_terragrunt_config`, a helper function that can be used for reading in another terragrunt config and accessing its blocks and attributes for reuse.

This is one of the suggested improvements coming out of https://github.com/gruntwork-io/terragrunt/pull/1025

### TODO

- [x] Update when https://github.com/gruntwork-io/terragrunt/pull/1050 merges (there are some conflicts in that `generate` blocks will need to be added to the converter).

brikis98 commented on config/config_helpers.go:
```diff
@@ -327,6 +328,81 @@ func getAWSCallerIdentityUserID(include *IncludeConfig, terragruntOptions *optio
+	// NOTE: this will not call terragrunt output, since all the values are cached from the ParseConfigFile call
+	// NOTE: we don't use range here because range will copy the slice, thereby undoing the set attribute.
+	for i := 0; i < len(config.TerragruntDependencies); i++ {
+		config.TerragruntDependencies[i].setRenderedOutputs(targetOptions)
```
Quick sanity check: any time I see a mutable call like this, I have to ask if it is thread / goroutine safe? E.g., Does this work properly during `apply-all`?

yorinasub17 replied:
Yes, because the `config` object is function scoped.

brikis98 commented on test/fixture-read-config/with_dependency/terragrunt.hcl:
```diff
@@ -0,0 +1,9 @@
+  source = "${get_terragrunt_dir()}/../../fixture-inputs"
+}
+
+inputs = local.config_with_dependency.dependency.inputs.outputs
```
NIT: Should we have a test case that checks that we're able to access every aspect of the config read in via `read_terragrunt_config`?

E.g., 

```hcl
locals {
  config = read_terragrunt_config("foo.hcl")
}

inputs = {
  locals = local.config.locals
  terraform = local.config.terraform
  inputs = local.config.inputs
  # etc
}
```

And the test checks that all the values in `inputs` are set to the exact values in `foo.hcl`?

yorinasub17 replied:
Done: bbf042e

yorinasub17 commented on config/config_as_cty.go:
```diff
@@ -0,0 +1,195 @@
+// Specifically, we want to reference blocks by named attributes, but blocks are rendered to lists in the
+// TerragruntConfig struct, so we need to do some massaging of the data to convert the list of blocks in to a map going
+// from the block name label to the block value.
+func terragruntConfigAsCty(config *TerragruntConfig) (cty.Value, error) {
```
Yea I was not very happy with this... but I also couldn't come up with a way to avoid it. I'll see if I can add a test here.

yorinasub17 replied:
Added some unit tests: c5ea857

brikis98 replied:
This test does a great job of checking that all the existing fields are converted, but if we were to add a new field, and forget to to add it to this test and to `terragruntConfigAsCty`... The test won't fail, right?

yorinasub17 replied:
Hmm my understanding is that [it should fail here](https://github.com/gruntwork-io/terragrunt/blob/yori-read-terragrunt-config/config/config_as_cty_test.go#L177), where it will exhaust the options when it sees a new struct field.

I'm going to verify this now that I merged in the codegen feature. My expectation is that the test will fail when I rebase and push. If it doesn't, I'll update the test until it fails.

yorinasub17 replied:
Yup failed as expected: https://circleci.com/gh/gruntwork-io/terragrunt/4101

```
=== RUN   TestTerragruntConfigAsCtyDrift
--- FAIL: TestTerragruntConfigAsCtyDrift (0.00s)
    config_as_cty_test.go:178: Unknown struct property: GenerateConfigs
```

brikis98 replied:
Ah, nice!

yorinasub17 commented:
I believe I addressed all the comments, but given the complexity of the new tests, I rerequested a review.

yorinasub17 commented:
> My only concern is about the maintainability of terragruntConfigAsCty... Or more accurately, of the 3-4 places we now have to update every time we add a new config option to Terragrunt.

Filed https://github.com/gruntwork-io/terragrunt/issues/1055 so we don't forget this thought.

yorinasub17 commented:
UPDATE: bfeab4c 

Added `generate` block and property to the output cty and updated the tests so that they pass and check that the block and property is accessible.

I also found a bug in the `generate` block where `comment_prefix` was ignored, so fixed that. 

Thanks for the review! Going to merge this in now!
